specs/deneb/polynomial-commitments-sampling.md

```diff
@@ -0,0 +1,530 @@
+                                 zs: Sequence[BLSFieldElement]) -> Tuple[KZGProof, Sequence[BLSFieldElement]]:
+    """
+    Helper function that computes multi-evaluation KZG proofs.
+    """
+
+    # For all x_i, compute p(x_i) - p(z)
+    ys = [evaluate_polynomialcoeff(polynomial_coeff, z) for z in zs]
+    interpolation_polynomial = interpolate_polynomialcoeff(zs, ys)
+    polynomial_shifted = add_polynomialcoeff(polynomial_coeff, neg_polynomialcoeff(interpolation_polynomial))
+
+    # For all x_i, compute (x_i - z)
+    denominator_poly = zero_polynomialcoeff(zs)
+
+    # Compute the quotient polynomial directly in evaluation form
+    quotient_polynomial = divide_polynomialcoeff(polynomial_shifted, denominator_poly)
+
+    return KZGProof(g1_lincomb(KZG_SETUP_G1[:len(quotient_polynomial)], quotient_polynomial)), ys
+```
+
+#### `verify_kzg_proof_multi_impl`
+
+```python
+def verify_kzg_proof_multi_impl(commitment: KZGCommitment,
+                                zs: BLSFieldElement,
+                                ys: BLSFieldElement,
+                                proof: KZGProof) -> bool:
+    """
+    Helper function that verifies a KZG multiproof
+    """
+    
+    zero_poly = g2_lincomb(KZG_SETUP_G2[:len(zs) + 1], zero_polynomialcoeff(zs))
+    interpolated_poly = g1_lincomb(KZG_SETUP_G1[:len(zs)], interpolate_polynomialcoeff(zs, ys))
+
+    return (
+        bls.pairing_check(
+            [[bls.bytes48_to_G1(proof), bls.bytes96_to_G2(zero_poly)],
+            [bls.add(bls.bytes48_to_G1(commitment), bls.neg(bls.bytes48_to_G1(interpolated_poly))),
+             bls.neg(bls.bytes96_to_G2(KZG_SETUP_G2[0]))]]
+        )
+    )
+```
+
+### Sample cosets
+
+#### `coset_for_sample`
+
+```python
+def coset_for_sample(sample_id: int) -> Vector[BLSFieldElement, FIELD_ELEMENTS_PER_SAMPLE]:
+    """
+    Get the subgroup for a given ``sample_id``
+    """
+    assert sample_id < SAMPLES_PER_BLOB
+    roots_of_unity_brp = bit_reversal_permutation(ROOTS_OF_UNITY2)
+    return Vector[BLSFieldElement, FIELD_ELEMENTS_PER_SAMPLE](
+        roots_of_unity_brp[FIELD_ELEMENTS_PER_SAMPLE * sample_id:FIELD_ELEMENTS_PER_SAMPLE * (sample_id + 1)]
+    )
+```
+
+## Samples
+
+### Sample computation
+
+#### `compute_samples_and_proofs`
+
+```python
+def compute_samples_and_proofs(blob: Blob) -> Tuple[
+                                                    Vector[KZGProof, SAMPLES_PER_BLOB], 
+                                                    Vector[Vector[BLSFieldElement, FIELD_ELEMENTS_PER_SAMPLE], SAMPLES_PER_BLOB]]:
+    """
+    Compute all the sample proofs for one blob. This is an inefficient O(n^2) algorithm,
+    for performant implementation the FK20 algorithm that runs in O(n log n) should be
+    used instead.
+    """
+    polynomial = blob_to_polynomial(blob)
+    polynomial_coeff = polynomial_eval_to_coeff(polynomial)
+
+    proofs = []
+    samples = []
+
+    for i in range(10):
```

Comment:
This `10` value appears to be hardcoded. This should be `SAMPLES_PER_BLOB` right?